core/config/stop_config.cpp

```diff
@@ -0,0 +1,151 @@
+// SPDX-FileCopyrightText: 2017 - 2024 The Ginkgo authors
+//
+// SPDX-License-Identifier: BSD-3-Clause
+
+#include <ginkgo/core/base/exception_helpers.hpp>
+#include <ginkgo/core/config/config.hpp>
+#include <ginkgo/core/config/registry.hpp>
+#include <ginkgo/core/solver/solver_base.hpp>
+#include <ginkgo/core/stop/criterion.hpp>
+#include <ginkgo/core/stop/iteration.hpp>
+#include <ginkgo/core/stop/residual_norm.hpp>
+#include <ginkgo/core/stop/time.hpp>
+
+
+#include "core/config/config_helper.hpp"
+#include "core/config/dispatch.hpp"
+
+
+namespace gko {
+namespace config {
+
+
+inline deferred_factory_parameter<stop::CriterionFactory> configure_time(
+    const pnode& config, const registry& context, type_descriptor td)
+{
+    auto factory = stop::Time::build();
+    if (auto& obj = config.get("time_limit")) {
+        factory.with_time_limit(gko::config::get_value<long long int>(obj));
+    }
+    return factory;
+}
+
+
+inline deferred_factory_parameter<stop::CriterionFactory> configure_iter(
+    const pnode& config, const registry& context, type_descriptor td)
+{
+    auto factory = stop::Iteration::build();
+    if (auto& obj = config.get("max_iters")) {
+        factory.with_max_iters(gko::config::get_value<size_type>(obj));
+    }
+    return factory;
+}
+
+
+stop::mode get_mode(const std::string& str)
+{
+    if (str == "absolute") {
+        return stop::mode::absolute;
+    } else if (str == "initial_resnorm") {
+        return stop::mode::initial_resnorm;
+    } else if (str == "rhs_norm") {
+        return stop::mode::rhs_norm;
+    }
+    GKO_INVALID_STATE("Not valid " + str);
+}
+
+
+template <typename ValueType>
+class ResidualNormConfigurer {
```

Comment:
why is here an extra type required to define the `build_from_config`. Why isn't `build_from_config` just part of `ResidualNorm`?

Reply:
because the decision is to provide LinOpFactory only, I need to make it separate from the public interface.
the getter functions need to be distinct for the same reason

Reply:
Maybe we should revisit that decision? It seems a bit inconsistent to me now.

Reply:
for me, it's pretty hard to judge anything on that. I am the one would like to support everything in the uniform way. 
I will be definitely okay to make this `parse` public available in the stop class, but I also try to stay with the decision.

Reply:
I think we should consider what kind of configuration we want:
```json
{
  "stop": [
    {
      "type": "stop::Time",
      "time": 1.5
    },
    {
      "type": "stop::ResidualNorm",
      "base": "relative_residual",
      "value": 1e-06
    },
    {
      "type": "stop::Iterations",
      "value": 100
    }
  ]
}
```
or
```json
{
  "stop": {
    "time": 1.5,
    "rel_residual_norm": 1e-06,
    "iterations": 100
  }
}
```
I believe the latter is better suited for users, and doesn't bring the aforementioned complexity with it. There is no purpose in adding the same type of stopping criterion twice, so the object approach doesn't restrict anything.

Reply:
the shortcut version is good to have but it is not restricted. Also, if we really think it is helpful, we should also provide this shortcut in normal type.
I would like to provide the same setup as the ginkgo type first, then we can provide the shortcut.
No, same type of stopping criterion makes sense. for example, I can check the absoulte residual norm and relative residual norm at the same time

Reply:
> No, same type of stopping criterion makes sense

by type I mean what kind of criterion we use formally (iteration, absolute norm, relative norm, implicit norm, time, ...), not the C++ types with multiple possible base values. They can all be specified by different keys like `_residual_reduction`, `rel_tolerance`, `abs_tolerance`, ... There is no advantage in using multiple instances of the same residual reduction stopping criterion with different tolerances.

We could have a map `string -> stop::Criterion(const config&)` with each key representing a possible type and the `config` we pass in gets interpreted by the stopping criterion. That also allows users to add their custom stopping criteria, while keeping the configuration itself slim. I believe this approach can achieve everything that the array of objects can achieve, while being more succinct, so I would prefer it.

Reply:
I would like to provide the same behavior first. It's not good we have different name for the same thing.
Well, I would like to provide the same thing like LinOpFactory, but it was rejected.
Also, it is tiny difference between `{"type": "...", "value": "..."}` -> `"type_value" : "..."`. the first one keep the clear and same responsibility as normal ginkgo type

Reply:
As a prerequisite, no of our stopping criteria need a ValueType, because they are always used only in conjunction with a solver, which could provide this type. So my long-term goal is that we can simplify our stopping criterion setup to
```cpp
build().with_criteria(stop::iterations(100), stop::rel_residual_norm(1e-6))
```
which maps well to these config keys. So instead of making the config options more like the factories, I would like to suggest making the factory setup more like the proposed config options.

As an example, take an OpenFOAM configuration, which achieves roughly the same thing, albeit at slightly less freedom of configuring solvers and preconditioners
```
{
        solver          GKOCG;
        preconditioner  none;
        executor        dpcpp;
        tolerance       1e-06;
        relTol          0.05;
}
```
That's how I would like to configure things from the user perspective, the whole `type` and `base` approach feels unnecessarily verbose.

Reply:
Yes, it is what I said. If we think the shortcut version good, we should provide it in ginkgo normal type.
but it is the future plan, and some solver does not require the value type.
I would like to go for what we have now first.
Also, in my opinion, almost all solver do not need the value type.

Reply:
> I would like to go for what we have now first.

I don't agree there - the configuration format is something that should be rather stable. If we settle on a configuration language, then we need to support it a long time, so we should not rely on something that we intend to change in the near future.

> Also, in my opinion, almost all solver do not need the value type.

I agree in general, as long as we find a good way of configuring mixed precision (e.g. based on value_type tags, which we probably need for this setup anyways)

Reply:
We can not ensure the thing it is indeed stable there. It's the future thing and we even do not have conclusion in the team.
Also, the proposed way here is not conflicted with the new way. I do not see any draw back on the current one except for a little more chars.
When we have that, it can be simply used without the array. That's another reason I allow one array object can use object input.

Reply:
As soon as we add the config, people will start using it, and they will want their old config files to work with newer versions of Ginkgo. So we should aim to keep the config files as stable as possible. We can achieve that with sufficient tests and by avoiding changes to the code that parses the config.

The proposed way does not conflict with the new way, but it requires (probably) more parsing code and more complex configuration for the same effect. Why would people use the array approach if they could use the more succinct object approach? I think it's important to try to think as a user configuring an application that uses Ginkgo under the hood, not an application developer that is familiar with Ginkgo. The former are who this config setup is most useful to IMO.

Reply:
It again the user type in our mind are different even they are the same role.
the future thing is not stable in my mind. The push here forced the others to make it stable not it is stable from the beginning if you said it is stable.
User should be able to try any combination as they want.
It is the same thing on the other way round, right? If they can configure it with more flexiblity, why they are resitricted?
If we provide both in normal ginkgo type, then we should provide both in the config.
Also, some solvers do not have the value type information.

Reply:
In our developer meeting on 17.05. we decided to:
1. Move forward with the currently available, more verbose specification.
2. Create an additional PR to support the shorter specification in addition to the current specification.

We will support both specification in the future (unless the second PR can be discussed and merged without slowing down our release schedule). The development burden is considered to be relatively small, since both specification will share a lot of implementation details. We might present one or the other more prominently in our (overhauled) documentation to clarify which one users should prefer.